core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadHandler.java

```diff
@@ -59,56 +62,80 @@ public class CompleteMultipartUploadHandler extends AbstractHandler {
-      final String object = s.substring(s.lastIndexOf(AlluxioURI.SEPARATOR) + 1);
-      final Long uploadId = Long.valueOf(request.getParameter("uploadId"));
-      httpServletResponse.setStatus(HttpServletResponse.SC_OK);
-      httpServletResponse.setContentType("text/xml");
-
-      Future<CompleteMultipartUploadResult> future =
-          mExecutor.submit(new CompleteMultipartUploadTask(mFileSystem, bucket, object, uploadId));
-      while (!future.isDone()) {
-        try {
-          Thread.sleep(mKeepAliveTime);
-        } catch (InterruptedException e) {
-          e.printStackTrace();
-        }
-        //periodically sends white space characters to keep the connection from timing out
-        httpServletResponse.getWriter().print(" ");
-        httpServletResponse.getWriter().flush();
-      }
+  public void handle(String s, Request request, HttpServletRequest httpServletRequest,
+                     HttpServletResponse httpServletResponse) throws IOException {
+    if (!s.startsWith(mS3Prefix)) {
+      return;
+    }
+    LOG.info("Alluxio S3 API received request: {}", request);
+    if (!request.getMethod().equals("POST")
+        || request.getParameter("uploadId") == null) {
+      return;
+    } // Otherwise, handle CompleteMultipartUpload
+    s = s.substring(mS3Prefix.length());
+    final String bucket = s.substring(0, s.indexOf(AlluxioURI.SEPARATOR));
+    final String object = s.substring(s.indexOf(AlluxioURI.SEPARATOR) + 1);
+    final String uploadId = request.getParameter("uploadId");
+    LOG.debug("(bucket: {}, object: {}, uploadId: {}) queuing task...",
+        bucket, object, uploadId);
 
-      XmlMapper mapper = new XmlMapper();
+    httpServletResponse.setStatus(HttpServletResponse.SC_OK);
+    httpServletResponse.setContentType(MediaType.APPLICATION_XML);
 
+    Future<CompleteMultipartUploadResult> future =
+        mExecutor.submit(new CompleteMultipartUploadTask(mFileSystem, bucket, object, uploadId,
+            IOUtils.toString(request.getReader())));
+    long sleepMs = 1000;
+    while (!future.isDone()) {
+      LOG.debug("(bucket: {}, object: {}, uploadId: {}) sleeping for {}ms...",
+          bucket, object, uploadId, sleepMs);
       try {
-        CompleteMultipartUploadResult result = future.get();
-        httpServletResponse.getWriter().write(mapper.writeValueAsString(result));
-      } catch (InterruptedException | ExecutionException e) {
-        e.printStackTrace();
+        Thread.sleep(sleepMs);
+      } catch (InterruptedException e) {
+        LOG.error(e.toString());
       }
+      //periodically sends white space characters to keep the connection from timing out
+      LOG.debug("(bucket: {}, object: {}, uploadId: {}) sending whitespace...",
+          bucket, object, uploadId);
+      httpServletResponse.getWriter().print(" ");
       httpServletResponse.getWriter().flush();
-      request.setHandled(true);
+      sleepMs = Math.min(2 * sleepMs, mKeepAliveTime);
```

Comment:
without this space character, would the server break the connection?

Reply:
Yes, see https://github.com/Alluxio/alluxio/issues/14464